File: Medium/Arrays & Hashing/product_of_array_except_self.js
function productExceptSelf(nums) {
    let answer = [];
    let prefix = [];
    let suffix = [];
    let product = 1;

    // prefix pass
    for (let i = 0; i < nums.length; i++) {
        if (i != 0) product = product * nums[i - 1];
        prefix.push(product);
    }

    product = 1;
    // suffix pass
    for (let j = nums.length - 1; j >= 0; j--) {
        if (j != nums.length - 1) product = product * nums[j + 1];
        suffix.push(product);
    }

    let i = 0;
    // multiply the two arrays
    for (let j = suffix.length - 1; j >= 0; j--) {
        answer.push(suffix[j] * prefix[i]);
        i++;
    }
    return answer;
}

console.log(productExceptSelf([-30, -40, -50, 1, 2, 3]));

File: Medium/Arrays & Hashing/product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
      prefix = [1]
      suffix = [1]
      result = []
      l = 0
      r = len(nums) - 1
      while l < len(nums) or r >= 0:
        if l > 0:
          prefix.append(prefix[-1] * nums[l - 1])
        if r < len(nums) - 1:
          suffix.append(suffix[-1] * nums[r + 1])
        l += 1
        r -= 1
      for i in range(len(prefix)):
        result.append(prefix[i] * suffix[len(nums) - 1 - i])
      return result

sol = Solution()
print(sol.productExceptSelf([-1,1,0,-3,3]))

# [1, 2, 3, 4]
#  1 3

# [1,2,3,4]
# [4, 3, 2, 1]
# [24,12,8,6]